booster/library/Booster/Pattern/ApplyEquations.hs

```diff
@@ -942,74 +920,246 @@ applyEquation term rule =
-        -- instantiate the requires clause with the obtained substitution
-        let required =
-                concatMap
-                    (splitBoolPredicates . coerce . substituteInTerm matchingSubst . coerce)
-                    rule.requires
-        -- If the required condition is _syntactically_ present in
-        -- the prior (known constraints), we don't check it.
-        knownPredicates <- (.predicates) <$> lift getState
-        toCheck <- lift $ filterOutKnownConstraints knownPredicates required
+filterOutKnownConstraints ::
+    forall io.
+    LoggerMIO io =>
+    Set Predicate ->
+    [Predicate] ->
+    EquationT io [Predicate]
+filterOutKnownConstraints priorKnowledge constraitns = do
+    let (knownTrue, toCheck) = partition (`Set.member` priorKnowledge) constraitns
+    unless (null knownTrue) $
+        getPrettyModifiers >>= \case
+            ModifiersRep (_ :: FromModifiersT mods => Proxy mods) ->
+                logMessage $
+                    renderOneLineText $
+                        "Known true side conditions (won't check):"
+                            <+> hsep (intersperse "," $ map (pretty' @mods) knownTrue)
+    pure toCheck
 
-        -- check the filtered requires clause conditions
-        unclearConditions <-
-            catMaybes
-                <$> mapM
-                    ( checkConstraint $ \p -> (\ctxt -> ctxt $ logMessage ("Condition simplified to #Bottom." :: Text), ConditionFalse p)
+checkRequires ::
+    forall io.
+    LoggerMIO io =>
+    Substitution ->
+    [Predicate] ->
+    [Predicate] ->
+    [Predicate] ->
+    ExceptT
+        ((EquationT io () -> EquationT io ()) -> EquationT io (), ApplyEquationFailure)
+        (EquationT io)
+        ()
+checkRequires currentSubst knownPredicates syntacticRequires restRequires' = do
+    ModifiersRep (_ :: FromModifiersT mods => Proxy mods) <- getPrettyModifiers
+    case syntacticRequires of
+        [] -> checkRequiresSemantically currentSubst (Set.fromList knownPredicates) restRequires'
+        _ ->
+            withContext CtxSyntactic $
+                checkRequiresSyntactically
+                    syntacticRequires
+                    currentSubst
+                    knownPredicates
+                    restRequires'
+
+checkRequiresSyntactically ::
+    forall io.
+    LoggerMIO io =>
+    [Predicate] ->
+    Substitution ->
+    [Predicate] ->
+    [Predicate] ->
+    ExceptT
+        ((EquationT io () -> EquationT io ()) -> EquationT io (), ApplyEquationFailure)
+        (EquationT io)
+        ()
+checkRequiresSyntactically syntacticRequires currentSubst knownPredicates restRequires' = do
+    ModifiersRep (_ :: FromModifiersT mods => Proxy mods) <- getPrettyModifiers
+    case syntacticRequires of
+        [] ->
+            unless (null restRequires') $
+                -- no more @syntacticRequires@, but unresolved conditions remain: abort
+                throwRemainingRequires currentSubst restRequires'
```

Comment:
why do we throw here?? i would expect at the base case to simply check the regular requires clauses...

Reply:
As I understand, the simplifications marked as `syntactic` must be only applied "syntactically", which means their side condition must not go to Z3. If we do not stop here, we will proceed to trying to discharge a syntactic simplification as a regular one, and as far as I understand, we do not want that. We want them either to succeed fast or to fail fast. @PetarMax do you agree?

Reply:
Regarding the code: we never want to transfer control to `checkRequiresSyntactically` from `checkRequiresSemantically`.

Reply:
So the semantics of the `syntactic()` attribute is:
- the `simplification` is syntactic, Z3 is never called
- the clause 1 is the driver clause which learns its variables from the known truth, and the the substitution for these learned variables is applied to all other clauses

I.e. the whole `simplification` is `syntactic`, not a clause of a simplification

Reply:
So, for me, the *clauses* marked as syntactic should be checked syntactically, the other ones should be checked semantically. A simplification can have both syntactic and semantic clauses.

This still does mean, I think, that `checkRequiresSyntactically` should not call `checkRequiresSemantically`.

Reply:
But also, all syntactics should be handled first, I think.